### PR TITLE
Implement analytics logger

### PR DESCRIPTION
**Pull request by mateherber**
# Context

As a part of the data overhaul project we need an easy-to-use API in our steps and tools that sends data to our [analytics service](https://github.com/bitrise-io/bitrise-step-analytics).

# Changes

- Created `Tracker` interface which can be used to send analytics events to our [analytics service](https://github.com/bitrise-io/bitrise-step-analytics)
  - Use `Wait()` method to wait for all job processed
- Created `Client` interface which is responsible for network communication

### CODE REVIEW

File: analytics/client.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"bytes"
+	"net/http"
+	"time"
+
+	"github.com/bitrise-io/go-utils/retry"
+)
+
+const trackEndpoint = "https://bitrise-step-analytics.herokuapp.com/track"
+const timeOutInSecs = 30
```
Comment (godrei) [inline analytics/client.go:12]: this would be simpler: `const timeout = 30 * time.Second`

File: analytics/client_test.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func Test_trackerClient_send(t *testing.T) {
+	testServer := httptest.NewServer(http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+		b, _ := io.ReadAll(req.Body)
+		assert.Equal(t, string(b), "{}")
+		assert.Equal(t, req.Method, http.MethodPost)
```
Comment (godrei) [inline analytics/client_test.go:16]: NOTE: might be worth adding a test for the content type of the request

File: analytics/client_test.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import (
+	"bytes"
+	"github.com/stretchr/testify/assert"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+func Test_trackerClient_send(t *testing.T) {
```
Comment (godrei) [inline analytics/client_test.go:12]: NOTE: might be worth adding a failing test case too.

File: analytics/client_test.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package analytics
+
+import (
+	"bytes"
```
Comment (godrei) [inline analytics/client_test.go:4]: forgot to run `goimports`?

File: analytics/track.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
+package analytics
+
+import (
+	"bytes"
+	"encoding/json"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
```
Comment (ofalvai) [inline analytics/track.go:9]: Shouldn't we use the other UUID package we agreed on in the RFC? https://bitrise.atlassian.net/wiki/spaces/ENGI/pages/2122547554/Standardize+used+uuid+package+for+Go+projects#%F0%9F%A4%94-Why?

File: analytics/track.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	sharedProperties map[string]interface{}
+}
+
+// NewDefaultTracker ...
+func NewDefaultTracker(shared map[string]interface{}) Tracker {
+	return NewTracker(make(chan *bytes.Buffer, bufferSize), &sync.WaitGroup{}, poolSize, NewDefaultClient(), shared)
+}
+
+// NewTracker ...
+func NewTracker(jobs chan *bytes.Buffer, waitGroup *sync.WaitGroup, size int, client Client, sharedProperties map[string]interface{}) Tracker {
```
Comment (godrei) [inline analytics/track.go:34]: do we need to wire in anything else than `client` and `sharedProperties` ? think neither in production nor in the test will we alter these.

Comment (godrei) [reply analytics/track.go:34]: NOTE: suggest unifying the var name of `sharedProperties` (I suggest `shared`) to clarify: both sets the same field of the struct.

File: analytics/track.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	t := tracker{jobs: jobs, waitGroup: waitGroup, client: client, sharedProperties: sharedProperties}
+	t.init(size)
+	return &t
+}
+
+// Enqueue ...
+func (t tracker) Enqueue(eventName string, properties map[string]interface{}) {
```
Comment (godrei) [inline analytics/track.go:41]: why not pass an `Event` here? by doing so later we could easier alter the Event struct (default formats and fields can be handled with a constructor for the Event struct)

Comment (godrei) [reply analytics/track.go:41]: might Event should be an interface, which can be JSONMarshalled, and shared fields can be set on it.

File: analytics/track.go (hunk 6 above)
Comment (ofalvai) [inline analytics/track.go:34]: I feel like it's not a good idea to let the caller decide the "pool size". Is this really expected that steps provide different size values? Why don't we just define a reasonable default value here, and later add a new constructor with a size parameter if needed?

Comment (mateherber) [reply analytics/track.go:34]: there is a default constructor that pins down that as well.

File: analytics/client_test.go (hunk 4 above)
Comment (mateherber) [reply analytics/client_test.go:4]: what is that?

Comment (mateherber) [reply analytics/client_test.go:4]: Oh you mean the order of the imports?

File: analytics/track.go (hunk 5 above)
Comment (mateherber) [reply analytics/track.go:9]: Didn't know about this 😳 will be using this then 👍🏻

File: analytics/track.go (hunk 6 above)
Comment (mateherber) [reply analytics/track.go:34]: Ok I'll hardcode the pool mechanism

File: analytics/track.go (hunk 7 above)
Comment (mateherber) [reply analytics/track.go:41]: I wouldn't set the shared properties on the Event, other than that it sounds ok

**Review by ofalvai — DISMISSED** [review]

**Review by godrei — DISMISSED** [review]

File: analytics/event.go [hunk 8]
```diff
@@ -8,46 +8,28 @@ import (
...
-func (e eventBuilder) AddProperty(key string, value interface{}) EventBuilder {
-	e.properties[key] = value
-	return e
-}
-
-// Build ...
-func (e eventBuilder) Build() Event {
-	return newEvent(e.eventName, e.properties)
+// EventDTO ...
+type EventDTO struct {
```
Comment (ofalvai) [inline analytics/event.go:12]: DTO as in Data Transfer Object? Is this Go lingo for data classes?

File: analytics/property.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+	GetValue() interface{}
+}
+
+type stringProperty struct {
+	key   string
+	value string
+}
+
+// StringProperty ...
+func StringProperty(key string, value string) Property {
```
Comment (ofalvai) [inline analytics/property.go:15]: So much of this file will be rewritten once Go 1.18 ships with generics 😬

**Review by ofalvai — DISMISSED** [review]

File: analytics/event.go (hunk 8 above)
Comment (mateherber) [reply analytics/event.go:12]: It's Java lingo for serializable classes 😅

File: analytics/property.go (hunk 9 above)
Comment (mateherber) [reply analytics/property.go:15]: Every Go file ever written will be rewritten after generics will be a thing 😛

**Review by godrei — APPROVED** [review]

**Review by godrei — DISMISSED** [review]

File: analytics/track.go [hunk 10]
```diff
@@ -2,62 +2,57 @@ package analytics
...
-		go t.worker()
-	}
-}
-
-func (t tracker) worker() {
-	for job := range t.jobs {
-		t.client.Send(job)
-		t.waitGroup.Done()
-	}
+// Fork ...
```
Comment (ofalvai) [inline analytics/track.go:55]:
```suggestion
// Pin ...
```

**Review by ofalvai — DISMISSED** [review]
LGTM, approved (but for the last time 😆 It's good enough.)

**Review by ofalvai — DISMISSED** [review]

**Pull request merged**